File: 2026/09-September/21-Monday/solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var resultArray = function(nums, k) {
    const ans = new Array(k).fill(0);
    let dp = new Array(k).fill(0);

    for (const num of nums) {
        const mod = num % k;
        const newDp = new Array(k).fill(0);

        // Start a new subarray with only num
        newDp[mod]++;

        // Extend previous subarrays
        for (let r = 0; r < k; r++) {
            if (dp[r] > 0) {
                const newRemainder = (r * mod) % k;
                newDp[newRemainder] += dp[r];
            }
        }

        // Add subarrays ending at the current position
        for (let r = 0; r < k; r++) {
            ans[r] += newDp[r];
        }

        dp = newDp;
    }

    return ans;
};

File: 2026/09-September/21-Monday/solution.py
from typing import List

class Solution:
    def resultArray(self, nums: List[int], k: int) -> List[int]:
        # ans[r] = total number of subarrays
        # whose product % k == r
        ans = [0] * k

        # dp[r] = number of subarrays ending at the
        # previous position with product % k == r
        dp = [0] * k

        for num in nums:
            num_mod = num % k

            # Subarrays ending at the current position
            new_dp = [0] * k

            # Start a new subarray containing only num
            new_dp[num_mod] += 1

            # Extend every previous subarray
            for r in range(k):
                new_r = (r * num_mod) % k
                new_dp[new_r] += dp[r]

            # Add current subarrays to the final answer
            for r in range(k):
                ans[r] += new_dp[r]

            dp = new_dp

        return ans
